File: epi_judge_cpp/int_as_array_increment.cc
#include <vector>
#include "test_framework/generic_test.h"
using std::vector;

/* Goal: Update given array that represents integer D to represent the integer D+1 
   Example: <1,2,9> -> <1,3,0>
*/

// My First Attempt Algorithm (Correct solution)
// Time Complexity: O(n)
// IDEA: Start from end of array, and add. If its 10, then we continue to add 1 to the next
// digit place. If we successfully added without a carry, then we simply return the array.

vector<int> PlusOne(vector<int> A) {
  for ( int i = A.size()-1; i >= 0; --i ) {
    if ( A[i] + 1 == 10 ) {
      A[i] = 0;
    } else {
      A[i] += 1;
      return A;
    }
  }

  // If we reach the end without returning, means we need to add 1 more 
  A.emplace(A.begin(),1);
  return A;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"A"};
  return GenericTestMain(args, "int_as_array_increment.cc",
                         "int_as_array_increment.tsv", &PlusOne,
                         DefaultComparator{}, param_names);
}

File: epi_judge_python/int_as_array_increment.py
from test_framework import generic_test

# Time Complexity: O(n)
# Space Complexity: O(1)
# Avg Run Time: 3ms
def plus_one(A:list):
    for i in reversed(range(0,len(A))):
        A[i] += 1
        if A[i] != 10:
            return A
        A[i] = 0

    A.insert(0,1)
    return A

# Avg Run Time: 12 ms
def TB_sol_plus_one(A) :
    A[-1] += 1
    for i in reversed(range(1, len(A))):
        if A[i] != 10:
            break
        A[i] = 0
        A[i-1] += 1
    if A[0] == 10:
    # There is a carry-out, so we need one more digit to store the resu-It
    # A slick way to do this is to append a 0 at the end of the array,
    # and update the fjrst entry to 1.
        A[0] = 1
        A. append (0)
    return A

if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("int_as_array_increment.py",
                                       "int_as_array_increment.tsv", plus_one))
